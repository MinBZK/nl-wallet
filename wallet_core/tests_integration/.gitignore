/config_server.toml
/gba_hc_converter.toml
/pid_issuer.toml
/update_policy_server.toml
/wallet.toml
/wallet_provider.toml
/wallet_server.toml
/wallet-config.json
/*.der
/*.pem
/*.der
*.snap.new
/*metadata.json
