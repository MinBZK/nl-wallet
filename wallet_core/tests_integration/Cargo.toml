[package]
name = "tests_integration"
version.workspace = true
edition.workspace = true
rust-version.workspace = true

[lints]
workspace = true

[[bin]]
name = "performance_test"
required-features = ["performance_test"]

[[test]]
name = "gba_pid_test"
path = "tests/gba_pid_test.rs"
required-features = ["gba_pid_test"]

[[test]]
name = "config_server"
path = "tests/config_server.rs"
required-features = ["integration_test"]

[[test]]
name = "digid_issuance"
path = "tests/digid_issuance.rs"
required-features = ["digid_test"]

[[test]]
name = "disclosure"
path = "tests/disclosure.rs"
required-features = ["integration_test"]

[[test]]
name = "issuance"
path = "tests/issuance.rs"
required-features = ["integration_test"]

[[test]]
name = "locking"
path = "tests/locking.rs"
required-features = ["integration_test"]

[[test]]
name = "registration"
path = "tests/registration.rs"
required-features = ["integration_test"]

[[test]]
name = "update_policy_server"
path = "tests/update_policy_server.rs"
required-features = ["integration_test"]

[features]
# Allow the disclosure return URL and its prefix to use http://
allow_insecure_url = ["openid4vc?/allow_insecure_url", "wallet_common?/allow_insecure_url"]
# Initialize global logging using tracing-subscriber
logging = ["dep:tracing", "dep:tracing-subscriber"]
# Include the common submodule, which includes code all integration tests use.
test_common = [
    "logging",
    "dep:android_attest",
    "dep:apple_app_attest",
    "dep:chrono",
    "dep:ctor",
    "dep:indexmap",
    "dep:jsonwebtoken",
    "dep:reqwest",
    "dep:sea-orm",
    "dep:serde_json",
    "dep:tokio",
    "dep:configuration_server",
    "dep:gba_hc_converter",
    "dep:nl_wallet_mdoc",
    "dep:openid4vc",
    "dep:platform_support",
    "dep:semver",
    "dep:sd_jwt",
    "dep:tempfile",
    "dep:update_policy_server",
    "dep:uuid",
    "dep:url",
    "dep:wallet",
    "dep:wallet_common",
    "dep:wallet_provider",
    "dep:wallet_provider_persistence",
    "dep:wallet_provider_service",
    "dep:wallet_server",
]
# Include the fake_digid submodule
fake_digid = ["dep:reqwest", "dep:url", "dep:wallet_common"]
# Include and run test that depend on an external PostgreSQL database
db_test = ["wallet_provider_persistence?/db_test", "wallet_provider_service?/db_test", "wallet_server?/db_test"]
# Include and run test that depend on a configured HSM
hsm_test = ["wallet_provider?/hsm_test"]
# Feature used for enabling all integration tests
integration_test = [
    "allow_insecure_url",
    "test_common",
    "db_test",
    "hsm_test",
    "dep:assert_matches",
    "dep:p256",
    "dep:rand_core",
    "dep:regex",
    "dep:rstest",
    "dep:serde_urlencoded",
    "dep:serial_test",
    "dep:tokio",
    "tokio/macros",
    "tokio/sync",
]
# Optional feature for running integration tests that depend on the DigiD connector
digid_test = ["fake_digid", "test_common", "dep:gba_hc_converter", "dep:rustls-pki-types"]
# Should be enabled when building the performance_test binary
performance_test = [
    "logging",
    "fake_digid",
    "dep:ctor",
    "dep:apple_app_attest",
    "dep:indexmap",
    "dep:reqwest",
    "dep:tokio",
    "dep:tracing",
    "dep:url",
    "dep:uuid",
    "dep:nl_wallet_mdoc",
    "dep:openid4vc",
    "dep:platform_support",
    "dep:serde_urlencoded",
    "dep:tempfile",
    "dep:wallet",
    "dep:wallet_common",
    "dep:wallet_server",
    "dep:dotenvy",
    "platform_support/mock_attested_key_apple_ca",
]
gba_pid_test = [
    "logging",
    "fake_digid",
    "dep:apple_app_attest",
    "dep:ctor",
    "dep:indexmap",
    "dep:insta",
    "dep:reqwest",
    "dep:tracing",
    "dep:url",
    "dep:uuid",
    "dep:nl_wallet_mdoc",
    "dep:openid4vc",
    "dep:platform_support",
    "dep:rstest",
    "dep:serial_test",
    "dep:serde_urlencoded",
    "dep:tokio",
    "dep:wallet",
    "dep:wallet_common",
    "dep:wallet_server",
    "platform_support/mock_attested_key_apple_ca",
    "wallet/snapshot_test",
]

[dependencies]
assert_matches = { workspace = true, optional = true }
ctor = { workspace = true, optional = true }
chrono = { workspace = true, optional = true }
indexmap = { workspace = true, optional = true }
insta = { workspace = true, optional = true, features = ["ron"] }
jsonwebtoken = { workspace = true, optional = true, features = ["use_pem"] }
reqwest = { workspace = true, optional = true, features = ["rustls-tls-webpki-roots"] }
sea-orm = { workspace = true, optional = true, features = [
    "sqlx-postgres",
    "runtime-tokio-rustls",
    "macros",
    "with-uuid",
    "debug-print",
] }
semver = { workspace = true, optional = true }
serde_json = { workspace = true, optional = true }
p256 = { workspace = true, optional = true, features = ["pkcs8"] }
rand_core = { workspace = true, optional = true }
regex = { workspace = true, optional = true }
rstest = { workspace = true, optional = true }
rustls-pki-types = { workspace = true, optional = true }
serde_urlencoded = { workspace = true, optional = true }
serial_test = { workspace = true, optional = true }
tempfile = { workspace = true, optional = true }
tokio = { workspace = true, optional = true, features = ["rt", "time", "parking_lot"] }
tracing = { workspace = true, optional = true }
tracing-subscriber = { workspace = true, optional = true, features = [
    "std",
    "fmt",
    "ansi",
    "tracing-log",
    "parking_lot",
] }
url = { workspace = true, optional = true }
uuid = { workspace = true, optional = true }

android_attest = { path = "../android_attest", optional = true }
apple_app_attest = { path = "../apple_app_attest", optional = true, features = ["mock"] }
configuration_server = { path = "../configuration_server", optional = true }
gba_hc_converter = { path = "../gba_hc_converter", optional = true }
nl_wallet_mdoc = { path = "../mdoc", optional = true, features = ["generate", "test"] }
openid4vc = { path = "../openid4vc", optional = true }
platform_support = { path = "../platform_support", optional = true, features = ["mock"] }
sd_jwt = { path = "../sd_jwt", optional = true, features = ["example_constructors"] }
update_policy_server = { path = "../update_policy_server", optional = true }
wallet = { path = "../wallet", optional = true, features = ["mock", "wallet_deps", "test"] }
wallet_common = { path = "../wallet_common", optional = true, features = ["mock_remote_key"] }
wallet_provider = { path = "../wallet_provider", optional = true }
wallet_provider_persistence = { path = "../wallet_provider/persistence", optional = true }
wallet_provider_service = { path = "../wallet_provider/service", optional = true, features = ["mock"] }
wallet_server = { path = "../wallet_server", optional = true, features = ["issuance", "disclosure", "mock"] }

[build-dependencies]
dotenvy = { workspace = true, optional = true }

[dev-dependencies]
# This enables the automatic running of unit and integration tests for these crates when tests
# are run from the workspace. The following features are not included in this:
# * hsm_test
# * db_test
openid4vc = { path = "../openid4vc", features = ["integration"] }
platform_support = { path = "../platform_support", features = ["persistent_mock_attested_key"] }
wallet_server = { path = "../wallet_server", features = ["issuance", "integration_test"] }
