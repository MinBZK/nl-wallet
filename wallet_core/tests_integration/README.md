# Integration tests

This crate contains several types of integration tests that require at least a running database.
All of these tests use the `wallet` crate (e.g. the Rust core of the wallet without the `flutter_api` layer) as a client
to test happy path flows against the backend services.

There are several categories of tests, which are gated behind features:

* A set of regular integration tests that require PostgreSQL and SoftHSM to be configured and running. These test the
  bulk of the wallet functionality and require the `integration_test` feature to be enabled.
* An integration test for issuance that includes the DigiD connector. This requires the `digid_test` feature to be
  enabled.
* A performance test that is set up as a separate binary. It requires the `performance_test` feature.

## Regular integration tests

To run these tests, perform the following steps:

* Set up the local environment using `./scripts/setup-devenv.sh`.
* Start PostgreSQL using `./scripts/start-devenv.sh postgres`.
* Run the tests from `wallet_core` using `cargo test --package tests_integration --features integration_test`.

## DigiD connector integration test

To run this test, perform the following steps:

* Set up the local environment using `./scripts/setup-devenv.sh`.
* Start PostgreSQL, the DigiD connector and the BRP proxy using `./scripts/start-devenv.sh postgres digid brpproxy`.
* Run the tests from `wallet_core` using `cargo test --package tests_integration --features digid_test`.

## Performance test

### Configuration

This test is meant to be run against an already running (external) environment. It uses the wallet default
configuration determined by `wallet/.env` (or by using environment variables).

In addition, it requires an `test_integration/.env` file containing these keys:

- `RELYING_PARTY_URL`: The external URL from where the disclosure flow is started. Is used as return_url.
- `INTERNAL_WALLET_SERVER_URL`: The internal URL where the disclosure session is started. Normally
  used by the relying party server.
- `PUBLIC_WALLET_SERVER_URL`: The public URL of the Wallet Server where the status for the session can be retrieved.
  Normally used by the wallet_web frontend.
- `APPLE_ATTESTATION_ENVIRONMENT`: The Apple attestation environment to be used for Apple key and app attestation that is configured in the Wallet Provider. Valid values are `development` and `production`, defaulting to the former if not provided.
- `TEAM_IDENTIFIER`: The Apple Team Identifier to be used for Apple key and app attestation that is configured in the Wallet Provider. Note that this needs to be provided together with `BUNDLE_IDENTIFIER`.
- `BUNDLE_IDENTIFIER`: The Bundle Identifier of the wallet app to be used for Apple key and app attestation that is configured in the Wallet Provider. Note that this needs to be provided together with `TEAM_IDENTIFIER`.

This file is automatically generated by the `scripts/setup-devenv.sh` script for running the performance test locally.

### Running locally

To run the test locally, the following servers need to be started:

    ./scripts/start-devenv.sh postgres wallet_provider configuration_server update_policy_server verification_server pid_issuer digid_connector brp

or shorter:

    ./scripts/start-devenv.sh postgres wp cs ups vs pi digid brp

To build from `wallet_core`:

    cargo build --release --bin performance_test --features performance_test,allow_insecure_url

And run with:

    ./target/release/performance_test

To run the test using Cargo from `wallet_core`:

    RUST_LOG=debug cargo run --bin performance_test --features performance_test,allow_insecure_url

### Running externally

Change the values in `test_integration/.env` appropriately. The `allow_insecure_url` feature is not necessary (but
also doesn't hurt). In addition, the `CONFIG_ENV` environment variable has to be set to point to the proper
environment, e.g. `ont`.

## GBA-V to PID tests

### Configuration

This test needs to run against the GBA-V proefomgeving. Therefore, the `gba-hc-converter` needs to be configured
according to the configuration as can be seen in `wallet_core/gba_hc_converter/gba_hc_converter.toml.gbav.example`.

### Running

To run the test, the following servers need to be started:

    ./scripts/start-devenv.sh postgres wallet_provider pid_issuer digid_connector brp

or shorter:

    ./scripts/start-devenv.sh postgres wp pi digid brp

And run the actual test with:

    cargo nextest run --package tests_integration --features=gba_pid_test --test gba_pid_test --fail-fast

### Maintenance

For mass accepting all `.snap.new` files, use the following command in the `tests/snapshots` folder:

    for i in ./*.snap.new ; do mv "$i" "${i%%.new}; done
