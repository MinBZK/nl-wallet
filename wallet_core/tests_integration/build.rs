#[cfg(feature = "performance_test")]
fn inject_dotenv_vars() {
    use std::env;
    use std::path::PathBuf;

    let crate_path: PathBuf = env::var("CARGO_MANIFEST_DIR").expect("Could not get crate path").into();
    let env_file_path = crate_path.join(".env");

    println!("cargo:rerun-if-changed={}", env_file_path.to_str().unwrap());

    match dotenvy::from_path_iter(env_file_path) {
        Ok(values) => {
            for item in values {
                let (key, value) = item.expect("Could not read entry from .env file");
                println!("cargo:rustc-env={key}={value}");
            }
        }
        // Do not panic on this, as we may want to operate without any `.env` file.
        Err(error) => println!("cargo:warning=Could not read .env file: {error}"),
    }
}

fn main() {
    #[cfg(feature = "performance_test")]
    inject_dotenv_vars();
}
