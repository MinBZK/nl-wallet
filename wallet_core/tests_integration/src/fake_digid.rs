use http::Method;
use reqwest::header::LOCATION;
use reqwest::redirect::Policy;
use reqwest::Response;
use url::Url;

use http_utils::reqwest::IntoPinnedReqwestClient;
use http_utils::reqwest::PinnedReqwestClient;
use http_utils::reqwest::ReqwestClientUrl;
use http_utils::tls::pinning::TlsPinningConfig;

// Use the mock flow of the DigiD bridge to simulate a DigiD login,
// invoking the same URLs at the DigiD bridge that would normally be invoked by the app and browser in the mock
// flow of the DigiD bridge.
// Note that this depends of part of the internal API of the DigiD bridge, so it may break when the bridge
// is updated.
pub async fn fake_digid_auth(mut authorization_url: Url, digid_http_config: TlsPinningConfig, bsn: &str) -> Url {
    let client = digid_http_config
        .try_into_custom_client(|client_builder| client_builder.redirect(Policy::none()))
        .unwrap();

    // Avoid the DigiD/mock DigiD landing page of the DigiD bridge by preselecting the latter
    authorization_url
        .query_pairs_mut()
        .append_pair("login_hint", "digid_mock");

    // Start authentication by GETting the authorization URL.
    // In the resulting HTML page, find the "RelayState" parameter which we need for the following URL.
    let relay_state_page = do_get_as_text(&client, ReqwestClientUrl::Absolute(authorization_url)).await;

    let relay_state_line = relay_state_page
        .lines()
        .find(|l| l.contains("RelayState"))
        .expect("failed to find RelayState");
    let relay_state = find_in_text(relay_state_line, "value=\"", "\"");

    // Note: the above HTTP response contains a HTML form that is normally automatically submitted
    // by the browser, leading to a page that contains the link that we invoke below.
    // To actually simulate autosubmitting that form and running some related JavaScript would be a bit of a hassle,
    // so here we skip autosubmitting that form. Turns out the DigiD bridge is fine with this.

    // Get the HTML page containing the redirect_uri back to our own app
    let finish_digid_path = format!("acs?SAMLart={}&RelayState={}&mocking=1", bsn, relay_state);

    let response = do_get_request(&client, ReqwestClientUrl::Relative(&finish_digid_path)).await;
    let redirect_url = response.headers().get(LOCATION).unwrap().to_str().unwrap();

    Url::parse(redirect_url).expect("failed to parse redirect url")
}

async fn do_get_request(client: &PinnedReqwestClient, url: ReqwestClientUrl<'_>) -> Response {
    client.send_request(Method::GET, url).await.expect("failed to GET URL")
}

async fn do_get_as_text(client: &PinnedReqwestClient, url: ReqwestClientUrl<'_>) -> String {
    do_get_request(client, url)
        .await
        .text()
        .await
        .expect("failed to get body text")
}

fn find_in_text<'a>(text: &'a str, start: &str, end: &str) -> &'a str {
    let start_index = text.find(start).expect("start not found");
    let remaining = &text[start_index + start.len()..];
    let end_index = remaining.find(end).expect("end not found");
    &remaining[..end_index]
}
