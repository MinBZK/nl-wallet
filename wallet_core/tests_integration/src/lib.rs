#[cfg(feature = "logging")]
pub mod logging;

#[cfg(feature = "test_common")]
pub mod common;
#[cfg(feature = "test_common")]
pub mod utils;

#[cfg(feature = "fake_digid")]
pub mod fake_digid;

#[cfg(feature = "performance_test")]
pub mod default {
    use apple_app_attest::AppIdentifier;
    use apple_app_attest::AttestationEnvironment;

    pub fn attestation_environment() -> AttestationEnvironment {
        option_env!("APPLE_ATTESTATION_ENVIRONMENT")
            .map(|environment| match environment {
                "development" => AttestationEnvironment::Development,
                "production" => AttestationEnvironment::Production,
                _ => panic!("Invalid Apple attestation environment"),
            })
            .unwrap_or(AttestationEnvironment::Development)
    }

    pub fn app_identifier() -> AppIdentifier {
        let team_identifier = option_env!("TEAM_IDENTIFIER");
        let bundle_identifier = option_env!("BUNDLE_IDENTIFIER");

        // Create an iOS app identifier if both environment variables are provided, otherwise fall back to the default.
        if let (Some(team_identifier), Some(bundle_identifier)) = (team_identifier, bundle_identifier) {
            AppIdentifier::new(team_identifier, bundle_identifier)
        } else {
            AppIdentifier::new("XGL6UKBPLP", "nl.ictu.edi.wallet.latest")
        }
    }
}
