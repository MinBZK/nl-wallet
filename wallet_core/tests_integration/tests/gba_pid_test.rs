use rstest::rstest;
use uuid::Uuid;

use openid4vc::disclosure_session::DisclosureSession;
use openid4vc::disclosure_session::HttpVpMessageClient;
use openid4vc::issuance_session::HttpIssuanceSession;
use openid4vc::issuance_session::IssuanceSessionError;
use openid4vc::ErrorResponse;
use openid4vc::TokenErrorCode;
use platform_support::attested_key::mock::MockHardwareAttestedKeyHolder;
use tests_integration::default;
use tests_integration::fake_digid::fake_digid_auth;
use wallet::errors::PidIssuanceError;
use wallet::mock::LocalConfigurationRepository;
use wallet::mock::MockStorage;
use wallet::mock::MockUpdatePolicyRepository;
use wallet::wallet_deps::default_wallet_config;
use wallet::wallet_deps::HttpAccountProviderClient;
use wallet::wallet_deps::HttpDigidSession;
use wallet::wallet_deps::Repository;
use wallet::wallet_deps::WpWteIssuanceClient;
use wallet::Wallet;

#[derive(Debug, Eq, PartialEq)]
enum TestError {
    Conversion,
    UnknownBsn,
    Unknown,
}

#[tokio::test]
#[rstest]
async fn test_gba_pid_conversion_error(
    #[values("999993318", "999993896", "999990585", "999991127", "999992326", "999997658")] bsn: &str,
) {
    assert_eq!(
        TestError::Conversion,
        gba_pid(bsn).await.expect_err("should return error")
    );
}

#[tokio::test]
#[rstest]
async fn test_gba_pid_unknown_bsn(#[values("900265462", "900253010")] bsn: &str) {
    assert_eq!(
        TestError::UnknownBsn,
        gba_pid(bsn).await.expect_err("should return error")
    );
}

#[tokio::test]
#[rstest]
async fn test_gba_pid_success(
    #[values(
        "999991772",
        "999991000",
        "999992958",
        "999991838",
        "999991802",
        "999991644",
        "999994761",
        "999990159",
        "999993598",
        "000009842",
        "999991747",
        "999994785",
        "999992636",
        "999993811",
        "999990640",
        "999992107",
        "999992120",
        "999991577",
        "999992533",
        "999994931",
        "999993215",
        "999992983",
        "999990044",
        "999990196",
        "999993446",
        "999992880",
        "000009878",
        "999991516",
        "999991292",
        "999992569",
        "999991814",
        "999994359",
        "999994542",
        "999990160",
        "999992065",
        "999991565",
        "999991243",
        "999990871",
        "999990500",
        "999993665",
        "999990627",
        "999993409",
        "999997634",
        "999997646",
        "999997671",
        "999997683",
        "999997695",
        "999997701",
        "999997713",
        "999997725",
        "999997737",
        "999997749",
        "999997750",
        "010245741",
        "010755561",
        "999998341",
        "999998353"
    )]
    bsn: &str,
) {
    assert!(gba_pid(bsn).await.is_ok());
}

async fn gba_pid(bsn: &str) -> Result<(), TestError> {
    let config_repository = LocalConfigurationRepository::new(default_wallet_config());
    let pid_issuance_config = &config_repository.get().pid_issuance;

    let mut wallet: Wallet<
        LocalConfigurationRepository,
        MockUpdatePolicyRepository,
        MockStorage,
        MockHardwareAttestedKeyHolder,
        HttpAccountProviderClient,
        HttpDigidSession,
        HttpIssuanceSession,
        DisclosureSession<HttpVpMessageClient, Uuid>,
        WpWteIssuanceClient,
    > = Wallet::init_registration(
        config_repository,
        MockUpdatePolicyRepository::default(),
        MockStorage::default(),
        MockHardwareAttestedKeyHolder::new_apple_mock(default::attestation_environment(), default::app_identifier()),
        HttpAccountProviderClient::default(),
    )
    .await
    .expect("Could not create test wallet");

    let pin = String::from("123344");

    wallet.register(pin.clone()).await.expect("Could not register wallet");

    let authorization_url = wallet
        .create_pid_issuance_auth_url()
        .await
        .expect("Could not create pid issuance auth url");

    let redirect_url = fake_digid_auth(&authorization_url, &pid_issuance_config.digid_http_config, bsn).await;

    let attestations_result = wallet.continue_pid_issuance(redirect_url).await;
    let attestations = match attestations_result {
        Ok(mdocs) => mdocs,
        Err(PidIssuanceError::PidIssuer(IssuanceSessionError::TokenRequest(ErrorResponse {
            error: TokenErrorCode::ServerError,
            error_description: Some(description),
            ..
        }))) if description.contains("Error converting GBA-V XML to Haal-Centraal JSON: GBA-V error") => {
            return Err(TestError::Conversion)
        }
        Err(PidIssuanceError::PidIssuer(IssuanceSessionError::TokenRequest(ErrorResponse {
            error: TokenErrorCode::ServerError,
            error_description: Some(description),
            ..
        }))) if description.contains("could not find attributes for BSN") => return Err(TestError::UnknownBsn),
        Err(e) => {
            dbg!("{:?}", e);
            return Err(TestError::Unknown);
        }
    };

    insta::with_settings!({
        description => format!("BSN: {}", bsn),
        snapshot_suffix => bsn,
        prepend_module_to_snapshot => false,
    }, {
        insta::assert_ron_snapshot!(attestations);
    });

    Ok(())
}
