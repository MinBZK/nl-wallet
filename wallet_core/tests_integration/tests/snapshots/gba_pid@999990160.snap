---
source: tests_integration/tests/gba_pid_test.rs
assertion_line: 194
description: "BSN: 999990160"
expression: attributes
---
{
  "com.example.pid": {
    [
      "family_name",
    ]: Basic("den Brinck"),
    [
      "given_name",
    ]: Basic("Hendrik Jan Coenraad Maurits"),
    [
      "birth_date",
    ]: Date("1971-12-13"),
    [
      "age_over_18",
    ]: Basic(true),
    [
      "bsn",
    ]: Basic("999990160"),
  },
  "com.example.address": {
    [
      "resident_street",
    ]: Basic("St.-Jacobsstraat"),
    [
      "resident_house_number",
    ]: Basic("1"),
    [
      "resident_postal_code",
    ]: Basic("3511BR"),
    [
      "resident_city",
    ]: Basic("Utrecht"),
    [
      "resident_country",
    ]: Basic("Nederland"),
  },
}
